p11-kit/generate-keypair.c

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright (c) 2023, Red Hat Inc.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above
+ *       copyright notice, this list of conditions and the
+ *       following disclaimer.
+ *     * Redistributions in binary form must reproduce the
+ *       above copyright notice, this list of conditions and
+ *       the following disclaimer in the documentation and/or
+ *       other materials provided with the distribution.
+ *     * The names of contributors to this software may not be
+ *       used to endorse or promote products derived from this
+ *       software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Zoltan Fridrich <zfridric@redhat.com>
+ */
+
+#include "config.h"
+
+#include "attrs.h"
+#include "compat.h"
+#include "debug.h"
+#include "iter.h"
+#include "message.h"
+#include "tool.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef ENABLE_NLS
+#include <libintl.h>
+#define _(x) dgettext(PACKAGE_NAME, x)
+#else
+#define _(x) (x)
+#endif
+
+int
+p11_kit_generate_keypair (int argc,
+			  char *argv[]);
+
+static bool
+get_mechanism (CK_MECHANISM *m,
+	       const char *type)
+{
+	if (p11_ascii_strcaseeq (type, "RSA"))
+		m->mechanism = CKM_RSA_PKCS_KEY_PAIR_GEN;
+	else if (p11_ascii_strcaseeq (type, "ECDSA"))
+		m->mechanism = CKM_ECDSA_KEY_PAIR_GEN;
+	else if (p11_ascii_strcaseeq (type, "Ed25519"))
+		m->mechanism = CKM_EC_EDWARDS_KEY_PAIR_GEN;
+	else
+		return false;
+
+	m->pParameter = NULL_PTR;
+	m->ulParameterLen = 0;
+
+	return true;
+}
+
+static const char *
+get_curve (const char *curve)
+{
+	if (p11_ascii_strcaseeq (curve, "SECP256R1"))
+		return "\x13\x05P-256";
```

Comment:
This is an encoded PrintableString in ASN.1, right? In that case we should treat it as a binary blob rather than a nul-terminated string (i.e., `uint8_t *` with explicit length, instead of `char *`). Also where are those curve names defined? It might be safer to use OIDs instead as defined in [SEC 2](https://www.secg.org/sec2-v2.pdf#page=33) as typical HSMs would recognize them.